File: takeout/December/Services/PIODefaultLeadTimeConfigurationService.cs
using Dapper;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using TAM.LogisticSystem.Entities;
using TAM.LogisticSystem.Models;

namespace TAM.LogisticSystem.Services
{
    public class PIODefaultLeadTimeConfigurationService
    {
        public PIODefaultLeadTimeConfigurationService(LogisticDbContext logisticDbContext)
        {
            this.LogisticDbContext = logisticDbContext;
        }

        private readonly LogisticDbContext LogisticDbContext;

        /// <summary>
        /// Get the list of PIO Default Lead Time, Location, and Process List from the database for 
        /// PIO Default Lead Time Configuration menu.
        /// </summary>
        /// <returns></returns>
        public async Task<PIODefaultLeadTimeConfigurationPageViewModel> GetPIODefaultLeadTimeConfigurationPage()
        {
            var defaultLeadTimeConfig = new PIODefaultLeadTimeConfigurationPageViewModel();
            defaultLeadTimeConfig.PIODefaultLeadTimes = (await this.LogisticDbContext.
                Database.GetDbConnection().QueryAsync<PIODefaultLeadTimeConfigurationViewModel>(@"
SELECT l.[Name] AS LocationName, pfl.[Name] AS ProcessStatus, ldltc.LeadMinutes AS LeadTime, 
l.LocationCode, pfl.ProcessForLineId, ldltc.RoutingMasterCode 
FROM LineDefaultLeadTimeConfiguration ldltc
JOIN [Location] l ON ldltc.LocationCode = l.LocationCode
JOIN ProcessForLine pfl ON ldltc.ProcessForLineId = pfl.ProcessForLineId 
WHERE ldltc.RoutingMasterCode = 'PIO'
")).ToList();

            defaultLeadTimeConfig.Locations = await this.LogisticDbContext.Location
                .Select(Q => new PIODefaultLeadTimeConfigurationLocationModel {
                    LocationCode = Q.LocationCode,
                    LocationName = Q.Name
                }).ToListAsync();

            defaultLeadTimeConfig.ProcessStatusList = await this.LogisticDbContext.ProcessForLine
                .ToListAsync();
            return defaultLeadTimeConfig;
        }

        /// <summary>
        /// Create a new PIO Default Lead Time data
        /// </summary>
        /// <param name="createPIODefaultLeadTime"></param>
        /// <returns></returns>
        public async Task<bool> CreatePIODefaultLeadTime(CreatePIODefaultLeadTimeViewModel createPIODefaultLeadTime)
        {
            var checkDuplicate = await this.LogisticDbContext.LineDefaultLeadTimeConfiguration
                .FirstOrDefaultAsync(Q => Q.RoutingMasterCode == "PIO" && Q.LocationCode == createPIODefaultLeadTime.LocationCode && Q.ProcessForLineId == createPIODefaultLeadTime.ProcessForLineId);

            if(checkDuplicate != null)
            {
                return false;
            }

            var newLineDefaultLeadTimeConfiguration = new LineDefaultLeadTimeConfiguration
            {
                RoutingMasterCode = "PIO",
                LocationCode = createPIODefaultLeadTime.LocationCode,
                ProcessForLineId = createPIODefaultLeadTime.ProcessForLineId,
                LeadMinutes = createPIODefaultLeadTime.TotalLeadTimeMinutes
            };

            this.LogisticDbContext.LineDefaultLeadTimeConfiguration.Add(newLineDefaultLeadTimeConfiguration);
            await this.LogisticDbContext.SaveChangesAsync();
            return true;
        }

        /// <summary>
        /// Update the existing PIO Default Lead Time data
        /// </summary>
        /// <param name="updatePIODefaultLeadTime"></param>
        /// <returns></returns>
        public async Task UpdatePIODefaultLeadTime(UpdatePIODefaultLeadTimeViewModel updatePIODefaultLeadTime)
        {
            await this.LogisticDbContext.Database.GetDbConnection().ExecuteAsync($@"
UPDATE
    LineDefaultLeadTimeConfiguration
SET
    LocationCode = @newLocationCode,
    ProcessForLineId = @newProcessForLineId,
    LeadMinutes = @leadTime
WHERE
    LocationCode = @oldLocationCode
    AND ProcessForLineId = @oldProcessForLineId
    AND RoutingMasterCode = @routingMasterCode
", new {
                routingMasterCode = updatePIODefaultLeadTime.RoutingMasterCode,
                oldLocationCode = updatePIODefaultLeadTime.OldLocationCode,
                newLocationCode = updatePIODefaultLeadTime.NewLocationCode,
                oldProcessForLineId = updatePIODefaultLeadTime.OldProcessForLineId,
                newProcessForLineId = updatePIODefaultLeadTime.NewProcessForLineId,
                leadTime = updatePIODefaultLeadTime.TotalLeadTimeMinutes
            });
        }

        /// <summary>
        /// Delete the existing PIO Default Lead Time data
        /// </summary>
        /// <param name="lineDefaultLeadTimeConfigurationId"></param>
        /// <returns></returns>
        public async Task<bool> DeletePIODefaultLeadTime(string routingMasterCode, string locationCode, int processForLineId)
        {
            var deletedLineDefaultLeadTimeConfiguration = await this.LogisticDbContext.LineDefaultLeadTimeConfiguration
                .FirstOrDefaultAsync(Q => Q.RoutingMasterCode == routingMasterCode && Q.LocationCode == locationCode && Q.ProcessForLineId == processForLineId);
            if(deletedLineDefaultLeadTimeConfiguration == null)
            {
                return false;
            }
            this.LogisticDbContext.LineDefaultLeadTimeConfiguration.Remove(deletedLineDefaultLeadTimeConfiguration);
            await this.LogisticDbContext.SaveChangesAsync();
            return true;
        }
    }
}

File: TAM.LogisticSystem/client/js/Services/PIODefaultLeadTimeConfigurationService.ts


export class PIODefaultLeadTimeConfigurationPageViewModel {
    locations: PIODefaultLeadTimeConfigurationLocationModel[];
    pioDefaultLeadTimes: [{
        leadTime: number,
        leadTimeToString: string,
        routingMasterCode: string,
        locationName: string,
        processStatus: string,
        locationCode: string,
        processForLineId: number
    }];
    processStatusList: ProcessStatusViewModel[];
}

export class PIODefaultLeadTimeConfigurationLocationModel {
    locationCode: string;
    locationName: string;
}

export class ProcessStatusViewModel {
    processForLineId: number;
    name: string;
}

export class PIOLeadTimeViewModel {
    day: number;
    hour: number;
    minute: number;
}

export class PIODefaultLeadTimeConfigurationUpdateFormViewModel {
    routingMasterCode: string;
    locationCode: string;
    processForLineId: number;
    totalLeadTimeMinutes: number;
}

export class PIOCreateDefaultLeadTimeConfigurationViewModel {
    locationCode: string;
    processForLineId: number;
    totalLeadTimeMinutes: number;
}

export class PIOUpdateDefaultLeadTimeConfigurationViewModel {
    routingMasterCode: string;
    oldLocationCode: string;
    newLocationCode: string;
    oldProcessForLineId: number;
    newProcessForLineId: number;
    totalLeadTimeMinutes: number;
}

export class PIODeleteDefaultLeadTimeConfigurationViewModel {
    routingMasterCode: string;
    locationCode: string;
    processForLineId: number;
}

export class PIODefaultLeadTimeConfigurationService {
    static $inject = ["$http"];
    
    constructor(httpClient: angular.IHttpService) {
        this.httpClient = httpClient;
    }

    httpClient: angular.IHttpService;

    // Retrieve the PIODefaultLeadTimeConfigurationPageViewModel from the server
    getData() {
        return this.httpClient.get<PIODefaultLeadTimeConfigurationPageViewModel>("/api/v1/PIODefaultLeadTimeConfigurationApi");
    }

    // Call the create a new PIO Default Lead Time method in the PIODefaultLeadTimeConfiguration controller
    createPIODefaultLeadTimeConfiguration(submitPIODefaultLeadTimeConfiguration: PIOCreateDefaultLeadTimeConfigurationViewModel) {
        return this.httpClient.post("/api/v1/PIODefaultLeadTimeConfigurationApi/Create", submitPIODefaultLeadTimeConfiguration);
    }

    // Call the update PIO Default Lead Time method in the PIODefaultLeadTimeConfiguration controller
    updatePIODefaultLeadTimeConfiguration(submitPIODefaultLeadTimeConfiguration: PIOUpdateDefaultLeadTimeConfigurationViewModel) {
        return this.httpClient.post("/api/v1/PIODefaultLeadTimeConfigurationApi/Update/", submitPIODefaultLeadTimeConfiguration);
    }

    // Call the delete PIO Default Lead Time method in the PIODefaultLeadTimeConfiguration controller
    deletePIODefaultLeadTimeConfiguration(pioDeleteDefaultLeadTimeConfiguration: PIODeleteDefaultLeadTimeConfigurationViewModel) {
        return this.httpClient.delete("/api/v1/PIODefaultLeadTimeConfigurationApi", { params: pioDeleteDefaultLeadTimeConfiguration });
    }
}